sphinx/tutorial/rose/furthertopics/upgrading-macro-development.rst

```diff
@@ -0,0 +1,568 @@
+
+This should provide you with a summary of changes (including any custom
+messages you may have added) and prompt you to accept them. Accept them
+and have a look at the app config file - it should have been changed
+accordingly.
+
+Using patch configurations
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+For relatively straightforward changes like the one above, we can
+configure a macro to apply patches to the configuration without having
+to write setting-specific Python code.
+
+We'll add a rudder option for our ``0.3`` version, with a
+``namelist:materials=l_rudder_branch``.
+
+Create a ``0.3`` directory in the same way that you created the ``0.1``
+and ``0.2`` metadata directories. Add a ``rose-meta.conf`` file that
+looks like this:
+
+.. code-block:: rose
+
+   [namelist:materials=hollow_tree_trunks]
+   compulsory=true
+   values=1
+
+   [namelist:materials=l_rudder_branch]
+   compulsory=true
+   type=logical
+
+   [namelist:materials=misc_branches]
+   compulsory=true
+   type=integer
+   range=4:
+
+   [namelist:materials=outrigger_tree_trunks]
+   compulsory=true
+   values=2
+
+   [namelist:materials=paddling_branches]
+   compulsory=true
+   range=1:
+   type=integer
+
+We need to write another macro in ``versions.py`` - append the following
+code:
+
+.. code-block:: python
+
+   class MySecondUpgradeMacro(rose.upgrade.MacroUpgrade):
+
+       """Upgrade from 0.2 (Outrageous Outrigger) to 0.3 (Amazing Ama)."""
+
+       BEFORE_TAG = "0.2"
+       AFTER_TAG = "0.3"
+
+       def upgrade(self, config, meta_config=None):
+           """Upgrade the boat!"""
+           self.act_from_files(config)
+           return config, self.reports
+
+The ``self.act_from_files`` line tells the macro to look for patch
+configuration files - two files called ``rose-macro-add.conf`` and
+``rose-macro-remove.conf``, under an ``etc/BEFORE_TAG/`` subdirectory -
+in our case, ``~/rose-meta/make-boat/etc/0.2/``.
+
+Whatever is found in ``rose-macro-add.conf`` will be added to the
+configuration, and whatever is found in ``rose-macro-remove.conf`` will
+be removed. If the files don't exist, nothing will happen.
+
+Let's configure what we want to happen. Create a directory
+``~/rose-meta/make-boat/etc/0.2/``, containing a ``rose-macro-add.conf``
+file that looks like this:
+
+.. code-block:: rose
+
+   [namelist:materials]
+   l_rudder_branch=.true.
+
+.. note::
+
+   If a ``rose-macro-add.conf`` setting is already defined, the
+   value will not be overwritten. In our case, we don't need a
+   ``rose-macro-remove.conf`` file.
+
+Go ahead and upgrade the app configuration to ``0.3``, as you did before.
+
+The ``rose-app.conf`` should now contain the new option, ``l_rudder_branch``.
+
+More Complex Upgrade Macros
+^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+.. TODO - link the upgrade macro API
+
+The Upgrade Macro API gives us quite a bit of power without having to write
+too much Python.
+
+For our ``1.0`` release, we want to add sailing gear - we need the number
```

Comment:
'sailing gear ~- we need~~ **so that** the number of ``misc_branches`` ~~to be~~ **is** at least ``6``, and **to** add a ``sail_canvas_sq_m option``.' In particular it is not clear that ``misc_branches`` counts as some form of ``sailing gear`` (as I assume is the case).